### PR TITLE
Add an abstract base class for data models

### LINKED ISSUE #340: Abstract base class for models
**Please give a one-sentence summary of the feature you would like to see.**

Models should always have certain functions.

**Please give as many details as possible about the requirements for the feature.**

Right now, all our models have a specific set of functions, but the next model could easily not have that function. We need a way to ensure that they have the following functions:

- get_attachment(self) -> Dict[str, Any]
- static to_dict(obj) -> Dict[str, Any]
- static from_dict(Dict[str, Any]) -> obj
- is_valid(self) -> bool

The static functions along with the `is_valid` function are used specifically for the database. `get_attachment` is used for slack attachments.

### PR DESCRIPTION
**Pull request by rmcreyes**
# Pull Request

## Description

**Give a brief description of your changes:**
Added an ABC for data models with the abstract methods mentioned in #340 

## Testing

**If testing this change requires extra setup, please document it here:**

## Ticket(s)

Closes #340

(Create a copy of that line for each Github Issue affected,
and replace "Affects" with "Closes" if merging this will close the relevant ticket.)

### CODE REVIEW

File: app/controller/command/commands/__init__.py [hunk 1]
```diff
@@ -3,5 +3,6 @@
 from app.controller.command.commands.team import TeamCommand
 from app.controller.command.commands.user import UserCommand
 from app.controller.command.commands.token import TokenCommand
+from app.controller.command.commands.project import ProjectCommand
 
-UnionCommands = Union[TeamCommand, UserCommand, TokenCommand]
+UnionCommands = Union[TeamCommand, UserCommand, TokenCommand, ProjectCommand]
```
Comment (cheukyin699) [inline app/controller/command/commands/__init__.py]: Should have no need for a union. We can use the parent class instead.

File: app/model/base.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    """Define the properties and methods needed for a data model."""
+
+    @abstractmethod
+    def get_attachment(self) -> Dict[str, Any]:
+        """Return slack-formatted attachment (dictionary) for data model."""
+        pass
+
+    @staticmethod
+    @abstractmethod
+    def to_dict(model: 'Model') -> Dict[str, Any]:
```
Comment (cheukyin699) [inline app/model/base.py]:
```suggestion
    def to_dict(model: 'RocketModel') -> Dict[str, Any]:
```

For every instance of `Model`.

File: app/controller/command/commands/__init__.py (hunk 1 above)
Comment (rmcreyes) [reply app/controller/command/commands/__init__.py]: Should this be addressed in another ticket? This change is only to address models.

Comment (cheukyin699) [reply app/controller/command/commands/__init__.py]: sure. a smaller ticket. very welcome

**Comment by cheukyin699** [comment]
This is weird. I'm going to look into this to see if there is a fix for it.

**Comment by cheukyin699** [comment]
image

I recommend using non-static methods (another PR) and bounded types. See [this stackoverflow][so] for more details. Be sure to make another issue for refactoring some of the methods into non-static methods.

[so]: https://stackoverflow.com/questions/44640479/mypy-annotation-for-classmethod-returning-instance

**Comment by rmcreyes** [comment]
Should those methods just be removed from the ABC for now then?

**Comment by cheukyin699** [comment]
Yeah let's remove them for now and add them later; it's a good first step.

**Pull request merged**